### PR TITLE
Added NUMA awareness for DpdkDevice (Issue #1414)

### PR DESCRIPTION
**Pull request by SesomB**
## Enhancement for `DpdkDevice` based on issue #1414 

The `DpdkDevice::openMultiQueues` method in the PcapPlusPlus DPDK wrapper is currently not NUMA aware.
It allocates memory using hard coded socket id, which is set to 0.
This behavior can lead to suboptimal performance on NUMA systems, where memory access times can vary significantly depending on the proximity of the memory to the CPU cores.

By implementing these changes, the `DpdkDevice::openMultiQueues` method will be NUMA aware, potentially improving performance on NUMA systems by ensuring that memory is allocated on the appropriate NUMA node.

### CODE REVIEW

**Comment by tigercosmos** [comment]
@SesomB could you provide a result of the different before/after, and also your computer hardware spec?

**Comment by SesomB** [comment]
@tigercosmos 

> @SesomB could you provide a result of the different before/after, and also your computer hardware spec?

Unfortunately, I currently don't have the right setup to utilize these changes. According to [DPDK's guidelines](https://doc.dpdk.org/guides/prog_guide/writing_efficient_code.html#numa), being NUMA aware should help with performance. These changes are based on my past experience with DPDK.

File: Pcap++/src/DpdkDevice.cpp [hunk 1]
```diff
@@ -119,6 +119,7 @@ DpdkDevice::DpdkDevice(int port, uint32_t mBufPoolSize, uint16_t mBufDataSize)
 	std::ostringstream deviceNameStream;
 	deviceNameStream << "DPDK_" << m_Id;
 	m_DeviceName = deviceNameStream.str();
+	m_DeviceSocketId =  rte_eth_dev_socket_id(m_Id);
```
Comment (seladb) [inline Pcap++/src/DpdkDevice.cpp:122]: nit: you have an extra space 🙂

**Review by seladb — COMMENTED** [review]
Overall looks good! Please see one minor comment.
I don't have a system with NUMA but I'll run the existing tests to make sure nothing breaks (unfortunately we currently don't have a way to run DPDK in CI)

File: Pcap++/src/DpdkDevice.cpp (hunk 1 above)
Comment (tigercosmos) [inline Pcap++/src/DpdkDevice.cpp:122]:
```suggestion
	m_DeviceSocketId =  rte_eth_dev_socket_id(m_Id);
```

Comment (tigercosmos) [reply Pcap++/src/DpdkDevice.cpp]: @seladb  I just fixed it, so should be okay to merge.

**Comment by seladb** [comment]
@SesomB the DPDK tests are failing with this new addition:
```
seladb@ubuntu22:~/PcapPlusPlus/Tests/Pcap++Test$ sudo Bin/Pcap++Test -i 10.0.3.15 -d 0 -t dpdk

Disabling memory leak check in MSVC Release builds due to caching logic in stream objects that looks like a memory leak:
     https://github.com/cpputest/cpputest/issues/786#issuecomment-148921958
Skipping memory leak check for all test cases
PcapPlusPlus version: v23.09+ (non-official release)
Built: Jun  5 2024 00:00:43
Git info: Git branch 'master', commit 'df105a79'
Using ip: 10.0.3.15
Debug mode: off
Using DPDK port: 0
DPDK KNI tests: skipped
Start running tests...

EAL: Detected CPU lcores: 6
EAL: Detected NUMA nodes: 1
EAL: Detected shared linkage of DPDK
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'PA'
EAL: Probe PCI driver: net_e1000_em (8086:100e) device: 0000:00:03.0 (socket -1)
TELEMETRY: No legacy callbacks, legacy socket not created
[ERROR: /home/seladb/PcapPlusPlus/Pcap++/src/DpdkDevice.cpp: initMemPool:439] Failed to create packets memory pool for port 0, pool name: MBufMemPool0. Error was: 'Cannot allocate memory' [Error code: 12]
[ERROR: /home/seladb/PcapPlusPlus/Pcap++/src/DpdkDevice.cpp: DpdkDevice:140] Could not initialize mBuf mempool. Device not initialized
TestDpdkInitDevice                 : PASSED
[ERROR: /home/seladb/PcapPlusPlus/Pcap++/src/DpdkDevice.cpp: configurePort:276] Could not open more than 0 TX queues
TestDpdkDevice                     : FAILED (/home/seladb/PcapPlusPlus/Tests/Pcap++Test/Tests/DpdkTests.cpp:313). Assert TRUE failed:
   [dev->open()]
[ERROR: /home/elad/PcapPlusPlus/Pcap++/src/DpdkDevice.cpp: configurePort:276] Could not open more than 0 TX queues
TestDpdkMultiThread                : FAILED (/home/elad/PcapPlusPlus/Tests/Pcap++Test/Tests/DpdkTests.cpp:398). Assert TRUE failed:
   [dev->openMultiQueues(numOfRxQueuesToOpen, 1)]
Ambiguous Rx mempools configuration
[ERROR: /home/elad/PcapPlusPlus/Pcap++/src/DpdkDevice.cpp: initQueues:373] Failed to init RX queue for device [DPDK_0]. Error was: 'Unknown error -22' [Error code: -22]
TestDpdkDeviceSendPackets          : FAILED (/home/elad/PcapPlusPlus/Tests/Pcap++Test/Tests/DpdkTests.cpp:550). Assert TRUE failed:
   [dev->openMultiQueues(1, txQueues, customConfig)]
TestDpdkDeviceWorkerThreads        : FAILED (/home/elad/PcapPlusPlus/Tests/Pcap++Test/Tests/DpdkTests.cpp:630). Assert TRUE failed:
   [dev->open()]
Ethdev port_id=0 nb_rx_queues=6 > 1
TestDpdkMbufRawPacket              : FAILED (/home/elad/PcapPlusPlus/Tests/Pcap++Test/Tests/DpdkTests.cpp:839). Assert TRUE failed:
   [dev->openMultiQueues(numOfRxQueues, numOfTxQueues)]

NOT ALL TESTS PASSED!!
Test cases: 89, Passed: 1, Failed: 5, Skipped: 83
```

I run it on Ubuntu 22.04 VM without NUMA.

I also verified that tests on `master` do pass on the same VM.

Can you please look into it?

**Comment by SesomB** [comment]
> @SesomB the DPDK tests are failing with this new addition:
> 
> ```
> seladb@ubuntu22:~/PcapPlusPlus/Tests/Pcap++Test$ sudo Bin/Pcap++Test -i 10.0.3.15 -d 0 -t dpdk
> ...............
> ...............
> ...............
> ...............
> ...............
> ...............
> ...............
> Test cases: 89, Passed: 1, Failed: 5, Skipped: 83
> ```
> 
> I run it on Ubuntu 22.04 VM without NUMA.
> 
> I also verified that tests on `master` do pass on the same VM.
> 
> Can you please look into it?


I ran the same test twice, once on the `master` branch without the multi-NUMA change and once on the `feature/dpdk-device-multi-numa` branch with the change. Interestingly, I got the same test results as you did (Passed: 1, Failed: 5) in both cases. In neither case did I encounter the mempool allocation failures you did.

From your test case, I noticed that DPDK printed this line: `EAL: Probe PCI driver: net_e1000_em (8086:100e) device: 0000:00:03.0 (socket -1)`. This indicates that the NIC is on socket `-1`, which is different from my setup. I believe the mempool failure on your machine is due to this `-1` socket ID. We may need to handle this special case.

`master` test results:

```
user@dev-server:~/repos/PcapPlusPlus$ sudo ./Tests/Pcap++Test/Bin/Pcap++Test -i 10.0.3.15 -d 0 -t dpdk
PcapPlusPlus version: v23.09+ (non-official release)
Built: Jun  5 2024 17:58:42
Git info: Git branch 'master', commit 'df105a79'
Using ip: 10.0.3.15
Debug mode: off
Using DPDK port: 0
DPDK KNI tests: skipped
Start running tests...

EAL: Detected CPU lcores: 8
EAL: Detected NUMA nodes: 1
EAL: Detected shared linkage of DPDK
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'PA'
EAL: VFIO support initialized
EAL: Using IOMMU type 8 (No-IOMMU)
EAL: Ignore mapping IO port bar(4)
EAL: Probe PCI driver: net_e1000_em (8086:100f) device: 0000:02:05.0 (socket 0)
TELEMETRY: No legacy callbacks, legacy socket not created
TestDpdkInitDevice                 : PASSED
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkDevice                     : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:353). Assert GREATER THAN failed:
   Actual:   0
   Expected: 0
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkMultiThread                : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:454). Assert GREATER THAN failed:
   Actual:   0
   Expected: 0
EAL: Error enabling MSI-X interrupts for fd 53
[ERROR: ../Pcap++/src/PcapFileDevice.cpp: open:273   ] Cannot open file reader device for filename 'PcapExamples/example.pcap': PcapExamples/example.pcap: No such file or directory
TestDpdkDeviceSendPackets          : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:553). Assert TRUE failed:
   [fileReaderDev.open()]
EAL: Error enabling MSI-X interrupts for fd 53
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkDeviceWorkerThreads        : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:674). Assert LOWER THAN failed:
   Actual:   20
   Expected: 20
EAL: Error enabling MSI-X interrupts for fd 53
[ERROR: ../Pcap++/src/PcapFileDevice.cpp: open:273   ] Cannot open file reader device for filename 'PcapExamples/example2.pcap': PcapExamples/example2.pcap: No such file or directory
TestDpdkMbufRawPacket              : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:846). Assert TRUE failed:
   [reader.open()]

NOT ALL TESTS PASSED!!
Test cases: 89, Passed: 1, Failed: 5, Skipped: 83
```

`feature/dpdk-device-multi-numa` test results:

```
user@dev-server:~/repos/PcapPlusPlus$ sudo ./Tests/Pcap++Test/Bin/Pcap++Test -i 10.0.3.15 -d 0 -t dpdk
PcapPlusPlus version: v23.09+ (non-official release)
Built: Jun  5 2024 18:08:30
Git info: Git branch 'feature/dpdk-device-multi-numa', commit 'edc1fdb3'
Using ip: 10.0.3.15
Debug mode: off
Using DPDK port: 0
DPDK KNI tests: skipped
Start running tests...

EAL: Detected CPU lcores: 8
EAL: Detected NUMA nodes: 1
EAL: Detected shared linkage of DPDK
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'PA'
EAL: VFIO support initialized
EAL: Using IOMMU type 8 (No-IOMMU)
EAL: Ignore mapping IO port bar(4)
EAL: Probe PCI driver: net_e1000_em (8086:100f) device: 0000:02:05.0 (socket 0)
TELEMETRY: No legacy callbacks, legacy socket not created
TestDpdkInitDevice                 : PASSED
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkDevice                     : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:353). Assert GREATER THAN failed:
   Actual:   0
   Expected: 0
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkMultiThread                : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:454). Assert GREATER THAN failed:
   Actual:   0
   Expected: 0
EAL: Error enabling MSI-X interrupts for fd 53
[ERROR: ../Pcap++/src/PcapFileDevice.cpp: open:273   ] Cannot open file reader device for filename 'PcapExamples/example.pcap': PcapExamples/example.pcap: No such file or directory
TestDpdkDeviceSendPackets          : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:553). Assert TRUE failed:
   [fileReaderDev.open()]
EAL: Error enabling MSI-X interrupts for fd 53
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkDeviceWorkerThreads        : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:674). Assert LOWER THAN failed:
   Actual:   20
   Expected: 20
EAL: Error enabling MSI-X interrupts for fd 53
[ERROR: ../Pcap++/src/PcapFileDevice.cpp: open:273   ] Cannot open file reader device for filename 'PcapExamples/example2.pcap': PcapExamples/example2.pcap: No such file or directory
TestDpdkMbufRawPacket              : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:846). Assert TRUE failed:
   [reader.open()]

NOT ALL TESTS PASSED!!
Test cases: 89, Passed: 1, Failed: 5, Skipped: 83
```
I've tested it on Ubuntu 20.04 VM without NUMA.

**Comment by SesomB** [comment]
Also, @seladb did you set up hugepages before testing ?

**Comment by seladb** [comment]
@SesomB your tests fail because you're not running them from the correct directory. What you should do is:
```
user@dev-server:~/repos/PcapPlusPlus$ cd Tests/Pcap++Test
user@dev-server:~/repos/PcapPlusPlus/Tests/Pcap++Test$ sudo Bin/Pcap++Test -i 10.0.3.15 -d 0 -t dpdk
```



> From your test case, I noticed that DPDK printed this line: `EAL: Probe PCI driver: net_e1000_em (8086:100e) device: 0000:00:03.0 (socket -1)`. This indicates that the NIC is on socket `-1`, which is different from my setup. I believe the mempool failure on your machine is due to this `-1` socket ID. We may need to handle this special case.

I believe socket ID `-1` is [`SOCKET_ID_ANY`](https://doc.dpdk.org/api/rte__memory_8h.html#a0307f4470d3f391102b0f489fc7d91b5) which should be an accepted value if NUMA is not supported, for example in [`rte_pktmbuf_pool_create()`](https://doc.dpdk.org/api/rte__mbuf_8h.html#a593921f13307803b94bbb4e0932db962) so I'm not sure why the tests fail...

> Also, @seladb did you set up hugepages before testing ?

Yes. I did it both when I ran on `master` (where all tests passed) and on your branch.

Let's see which result you get on your machine. By the way, do you see a socket ID which is different from `-1`?

**Comment by SesomB** [comment]
> By the way, do you see a socket ID which is different from `-1`?

Yes. As you can see from my tests, in the line that starts with `EAL: Probe PCI driver..`, on my machine, the NIC is on socket `0`.

> Let's see which result you get on your machine.

So, I re-ran the tests from the correct directory this time, and the problem I had continued to appear in both branches. (Test cases: 89, Passed: 2, Failed: 4, Skipped: 83)

### Test results - `feature/dpdk-device-multi-numa` :
```
user@dev-server:~/repos/PcapPlusPlus/Tests/Pcap++Test$ sudo Bin/Pcap++Test -i 10.0.3.15 -d 0 -t dpdk
PcapPlusPlus version: v23.09+ (non-official release)
Built: Jun  6 2024 17:54:21
Git info: Git branch 'feature/dpdk-device-multi-numa', commit '79744916'
Using ip: 10.0.3.15
Debug mode: off
Using DPDK port: 0
DPDK KNI tests: skipped
Start running tests...

EAL: Detected CPU lcores: 8
EAL: Detected NUMA nodes: 1
EAL: Detected shared linkage of DPDK
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'PA'
EAL: VFIO support initialized
EAL: Using IOMMU type 8 (No-IOMMU)
EAL: Ignore mapping IO port bar(4)
EAL: Probe PCI driver: net_e1000_em (8086:100f) device: 0000:02:05.0 (socket 0)
TELEMETRY: No legacy callbacks, legacy socket not created
TestDpdkInitDevice                 : PASSED
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkDevice                     : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:353). Assert GREATER THAN failed:
   Actual:   0
   Expected: 0
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkMultiThread                : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:454). Assert GREATER THAN failed:
   Actual:   0
   Expected: 0
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkDeviceSendPackets          : PASSED
EAL: Error enabling MSI-X interrupts for fd 53
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkDeviceWorkerThreads        : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:674). Assert LOWER THAN failed:
   Actual:   20
   Expected: 20
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkMbufRawPacket              : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:914). Assert LOWER THAN failed:
   Actual:   30
   Expected: 30

NOT ALL TESTS PASSED!!
Test cases: 89, Passed: 2, Failed: 4, Skipped: 83
```

### Test results - `master` :
```
user@dev-server:~/repos/PcapPlusPlus/Tests/Pcap++Test$ sudo Bin/Pcap++Test -i 10.0.3.15 -d 0 -t dpdk
PcapPlusPlus version: v23.09+ (non-official release)
Built: Jun  6 2024 17:10:41
Git info: Git branch 'master', commit '79f26320'
Using ip: 10.0.3.15
Debug mode: off
Using DPDK port: 0
DPDK KNI tests: skipped
Start running tests...

EAL: Detected CPU lcores: 8
EAL: Detected NUMA nodes: 1
EAL: Detected shared linkage of DPDK
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'PA'
EAL: VFIO support initialized
EAL: Using IOMMU type 8 (No-IOMMU)
EAL: Ignore mapping IO port bar(4)
EAL: Probe PCI driver: net_e1000_em (8086:100f) device: 0000:02:05.0 (socket 0)
TELEMETRY: No legacy callbacks, legacy socket not created
TestDpdkInitDevice                 : PASSED
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkDevice                     : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:353). Assert GREATER THAN failed:
   Actual:   0
   Expected: 0
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkMultiThread                : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:454). Assert GREATER THAN failed:
   Actual:   0
   Expected: 0
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkDeviceSendPackets          : PASSED
EAL: Error enabling MSI-X interrupts for fd 53
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkDeviceWorkerThreads        : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:674). Assert LOWER THAN failed:
   Actual:   20
   Expected: 20
EAL: Error enabling MSI-X interrupts for fd 53
TestDpdkMbufRawPacket              : FAILED (../Tests/Pcap++Test/Tests/DpdkTests.cpp:914). Assert LOWER THAN failed:
   Actual:   30
   Expected: 30

NOT ALL TESTS PASSED!!
Test cases: 89, Passed: 2, Failed: 4, Skipped: 83
```

I'm starting to suspect that the issue is related to the line `EAL: Error enabling MSI-X interrupts for fd 53`.

Could you show me the output of `dpdk-devbind.py --status` on your machine, especially the section regarding `Network devices using DPDK-compatible driver`?

Here is the output from my machine for comparison:
```
user@dev-server:~/repos/PcapPlusPlus/Tests/Pcap++Test$ dpdk-devbind.py --status

Network devices using DPDK-compatible driver
============================================
0000:02:05.0 '82545EM Gigabit Ethernet Controller (Copper) 100f' drv=vfio-pci unused=e1000
```
The driver I'm using is `vfio-pci`, and I suspect that it may be different from yours.

**Comment by seladb** [comment]
@SesomB yes, I use a different driver - `uio_pci_generic`. Here is my setup:
```
seladb@ubuntu22:~/PcapPlusPlus/build$ sudo python3 setup_dpdk.py setup -g 512 -i enp0s3 -m uio_pci_generic
mount: /mnt/huge: mount point does not exist.
[INFO] set up hugepages to 512
[INFO] loaded kernel module 'uio_pci_generic'
[INFO] set interface 'enp0s3' down
[INFO] bound interface 'enp0s3' ['0000:00:03.0'] to 'uio_pci_generic'
[INFO] SETUP COMPLETE
seladb@ubuntu22:~/PcapPlusPlus/build$ sudo python3 setup_dpdk.py status

Network devices using DPDK-compatible driver
============================================
0000:00:03.0 '82540EM Gigabit Ethernet Controller 100e' drv=uio_pci_generic unused=e1000,igb_uio,vfio-pci

Network devices using kernel driver
===================================
0000:00:08.0 '82540EM Gigabit Ethernet Controller 100e' if=enp0s8 drv=e1000 unused=igb_uio,vfio-pci,uio_pci_generic *Active*

No 'Baseband' devices detected
==============================

No 'Crypto' devices detected
============================

No 'Eventdev' devices detected
==============================

No 'Mempool' devices detected
=============================

No 'Compress' devices detected
==============================

No 'Misc (rawdev)' devices detected
===================================
```

I ran it again on your branch and on `master` just to make sure:

## Running on `master`

```
seladb@ubuntu22:~/PcapPlusPlus/Tests/Pcap++Test$ sudo Bin/Pcap++Test -i 10.0.3.15 -d 0 -t dpdk
Disabling memory leak check in MSVC Release builds due to caching logic in stream objects that looks like a memory leak:
     https://github.com/cpputest/cpputest/issues/786#issuecomment-148921958
Skipping memory leak check for all test cases
PcapPlusPlus version: v23.09+ (non-official release)
Built: Jun  5 2024 00:00:43
Git info: Git branch 'master', commit 'df105a79'
Using ip: 10.0.3.15
Debug mode: off
Using DPDK port: 0
DPDK KNI tests: skipped
Start running tests...

EAL: Detected CPU lcores: 6
EAL: Detected NUMA nodes: 1
EAL: Detected shared linkage of DPDK
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'PA'
EAL: Probe PCI driver: net_e1000_em (8086:100e) device: 0000:00:03.0 (socket -1)
TELEMETRY: No legacy callbacks, legacy socket not created
TestDpdkInitDevice                 : PASSED
TestDpdkDevice                     : PASSED
TestDpdkMultiThread                : PASSED
TestDpdkDeviceSendPackets          : PASSED
TestDpdkDeviceWorkerThreads        : PASSED
TestDpdkMbufRawPacket              : PASSED

ALL TESTS PASSED!!
Test cases: 89, Passed: 6, Failed: 0, Skipped: 83
```

## Running on `SesomB:feature/dpdk-device-multi-numa`

```
seladb@ubuntu22:~/PcapPlusPlus/Tests/Pcap++Test$ sudo Bin/Pcap++Test -i 10.0.3.15 -d 0 -t dpdk
Disabling memory leak check in MSVC Release builds due to caching logic in stream objects that looks like a memory leak:
     https://github.com/cpputest/cpputest/issues/786#issuecomment-148921958
Skipping memory leak check for all test cases
PcapPlusPlus version: v23.09+ (non-official release)
Built: Jun  5 2024 00:00:43
Git info: Git branch 'master', commit 'df105a79'
Using ip: 10.0.3.15
Debug mode: off
Using DPDK port: 0
DPDK KNI tests: skipped
Start running tests...

EAL: Detected CPU lcores: 6
EAL: Detected NUMA nodes: 1
EAL: Detected shared linkage of DPDK
EAL: Multi-process socket /var/run/dpdk/rte/mp_socket
EAL: Selected IOVA mode 'PA'
EAL: Probe PCI driver: net_e1000_em (8086:100e) device: 0000:00:03.0 (socket -1)
TELEMETRY: No legacy callbacks, legacy socket not created
[ERROR: /home/seladb/PcapPlusPlus/Pcap++/src/DpdkDevice.cpp: initMemPool:439] Failed to create packets memory pool for port 0, pool name: MBufMemPool0. Error was: 'Cannot allocate memory' [Error code: 12]
[ERROR: /home/seladb/PcapPlusPlus/Pcap++/src/DpdkDevice.cpp: DpdkDevice:140] Could not initialize mBuf mempool. Device not initialized
TestDpdkInitDevice                 : PASSED
[ERROR: /home/seladb/PcapPlusPlus/Pcap++/src/DpdkDevice.cpp: configurePort:276] Could not open more than 0 TX queues
TestDpdkDevice                     : FAILED (/home/seladb/PcapPlusPlus/Tests/Pcap++Test/Tests/DpdkTests.cpp:313). Assert TRUE failed:
   [dev->open()]
[ERROR: /home/seladb/PcapPlusPlus/Pcap++/src/DpdkDevice.cpp: configurePort:276] Could not open more than 0 TX queues
TestDpdkMultiThread                : FAILED (/home/seladb/PcapPlusPlus/Tests/Pcap++Test/Tests/DpdkTests.cpp:398). Assert TRUE failed:
   [dev->openMultiQueues(numOfRxQueuesToOpen, 1)]
Ambiguous Rx mempools configuration
[ERROR: /home/seladb/PcapPlusPlus/Pcap++/src/DpdkDevice.cpp: initQueues:373] Failed to init RX queue for device [DPDK_0]. Error was: 'Unknown error -22' [Error code: -22]
TestDpdkDeviceSendPackets          : FAILED (/home/seladb/PcapPlusPlus/Tests/Pcap++Test/Tests/DpdkTests.cpp:550). Assert TRUE failed:
   [dev->openMultiQueues(1, txQueues, customConfig)]
TestDpdkDeviceWorkerThreads        : FAILED (/home/seladb/PcapPlusPlus/Tests/Pcap++Test/Tests/DpdkTests.cpp:630). Assert TRUE failed:
   [dev->open()]
Ethdev port_id=0 nb_rx_queues=6 > 1
TestDpdkMbufRawPacket              : FAILED (/home/seladb/PcapPlusPlus/Tests/Pcap++Test/Tests/DpdkTests.cpp:839). Assert TRUE failed:
   [dev->openMultiQueues(numOfRxQueues, numOfTxQueues)]

NOT ALL TESTS PASSED!!
Test cases: 89, Passed: 1, Failed: 5, Skipped: 83
```

**Comment by SesomB** [comment]
> @SesomB yes, I use a different driver - `uio_pci_generic`.

I'm having trouble binding my NIC to `uio_pci_generic`, though it binds fine with `vfio-pci` and `igb_uio`. However, the tests still don't fully pass with these two drivers.

As you said earlier, `-1` is a valid value for allocating with the `rte` library, which is why I can't seem to understand why the tests are failing.

Let's set aside the driver differences for now and focus on the code. Since your NIC is on socket `-1`, could you print out the return values of `rte_socket_id()` and `rte_eth_dev_socket_id(m_Id)` (after successfully initializing `rte_eal`) ? 


This will help me understand why the tests aren't passing and allow me to adjust the code accordingly.

**Comment by seladb** [comment]
@SesomB I found the issue! `m_DeviceSocketId` should be of type `int` and not `uint16_t`.

[`rte_pktmbuf_pool_create()`](https://doc.dpdk.org/api/rte__mbuf_8h.html#a593921f13307803b94bbb4e0932db962) expects an `int` that's can be `-1` or [SOCKET_ID_ANY](https://doc.dpdk.org/api/rte__memory_8h.html#a0307f4470d3f391102b0f489fc7d91b5) if the device has no NUMA. But with `uint16_t` the value of `-1` translates to `65535` which is not really `-1`.

If we're concerned about an `int` (4 or 8 bytes, depending on the system) we can also store it as `int8_t` or `int16_t`. I tested both and they work

**Comment by SesomB** [comment]
> @SesomB I found the issue! `m_DeviceSocketId` should be of type `int` and not `uint16_t`.

@seladb 
Thanks for catching that! It was my mistake to use `uint16_t` for `m_DeviceSocketId`. Great find on the type mismatch causing the issue.


I'll update the code to use `int_16_t` for `m_DeviceSocketId`, I'll commit the changes shortly.

**Comment by seladb** [comment]
> > @SesomB I found the issue! `m_DeviceSocketId` should be of type `int` and not `uint16_t`.
> 
> @seladb Thanks for catching that! It was my mistake to use `uint16_t` for `m_DeviceSocketId`. Great find on the type mismatch causing the issue.
> 
> I'll update the code to use `int16_t` for `m_DeviceSocketId`, I'll commit the changes shortly.

Sure no problem! By the way, did you get it to run on your machine?

**Comment by SesomB** [comment]
> Sure, no problem! By the way, did you get it to run on your machine?

Unfortunately, I'm still having trouble binding to `uio_pci_generic`. On `vfio-pci` and `igb_uio`, the tests don't pass, likely due to the `EAL: Error enabling MSI-X interrupts for fd 53` message. This seems to be affecting the test results. I'll keep investigating this issue; I found some leads [here on GitHub](https://github.com/iqiyi/dpvs/issues/831) and on [this dev.to blog](https://dev.to/dannypsnl/dpdk-eal-input-output-error-1kn4). It seems that it might be a VMware-related issue.

I'll remind you that on my testing machine, the `uint16_t` type wasn't detected as an issue since my NIC was somehow on socket `0` instead of `-1` like yours.

**Review by seladb — APPROVED** [review]

**Comment by seladb** [comment]
> > Sure, no problem! By the way, did you get it to run on your machine?
> 
> Unfortunately, I'm still having trouble binding to `uio_pci_generic`. On `vfio-pci` and `igb_uio`, the tests don't pass, likely due to the `EAL: Error enabling MSI-X interrupts for fd 53` message. This seems to be affecting the test results. I'll keep investigating this issue; I found some leads [here on GitHub](https://github.com/iqiyi/dpvs/issues/831) and on [this dev.to blog](https://dev.to/dannypsnl/dpdk-eal-input-output-error-1kn4). It seems that it might be a VMware-related issue.
> 
> I'll remind you that on my testing machine, the `uint16_t` type wasn't detected as an issue since my NIC was somehow on socket `0` instead of `-1` like yours.

On my VirtualBox VM I can't bind to `vfio-pci`, not sure why yet 😕 
Let me know if you find what the problem is, maybe we can share it in the project documentation

**Comment by seladb** [comment]
Thank you @SesomB for adding this to PcapPlusPlus, much appreciated! 🙏

**Comment by SesomB** [comment]
Thank you, @seladb! I'm glad to contribute to PcapPlusPlus. 🙌

I also sent you an email regarding an issue I'd like to discuss further. Looking forward to your response!

**Pull request merged**